### PR TITLE
CP KSQL CLI docker image

### PR DESCRIPTION
**Pull request by apurvam**
### Description 
This patch introduces a new CP KSQL CLI docker image, which provides just the KSQL CLI. In the current setup, this image will _always_ run the command `ksql`. By default, the server url will be `http://localhost:8088`, but this can be overridden on the CLI. Any further command line arguments to the CLI can also be provided after the server URL.

Since we can't log to stdout (that will interfere with the actual CLI), the logs go to `/var/logs/ksql-cli/`.

Examples follow.

#### Connect to a dockerized ksql server
This works because of the docker network created by the ksql server.

```bash
# Run the server
$ docker run -d -p 10.0.0.11:8088:8088 \
    -e KSQL_BOOTSTRAP_SERVERS=amehta-macbook-pro-2.local:9092 \
    -e KSQL_OPTS="-Dksql.service.id=amehta_test_3_  -Dlisteners=http://0.0.0.0:8088/" \  
    confluentinc/cp-ksql-server

# Now connect the CLI to it.
$ docker run -rm -it confluentinc/cp-ksql-cli http://10.0.0.11:8088 
.. clipped ascii art ..
Copyright 2017 Confluent Inc.

CLI v4.1.2-SNAPSHOT, Server v4.1.2-SNAPSHOT located at http://10.0.0.11:8088

Having trouble? Type 'help' (case-insensitive) for a rundown of how things work!

ksql>
```

#### Provide a config file
```bash
# Assume server is running.
$ ls ~/workspace/docker-mount
ksql-cli.properties
$ docker run -rm -it \
      -v ~/workspace/docker-mount/:/mnt/ksql-cli  \
     confluentinc/cp-ksql-cli http://10.0.0.11:8088 \
        --config-file /mnt/ksql-cli/ksql-cli.properties

... cli ...
```

#### Connect to a ksql server running on another host, like on AWS:
```bash
$ docker run -it confluentinc/cp-ksql-cli \
     http://ec2-blah.us-blah.compute.amazonaws.com:8080
... ascii art ..
Copyright 2017 Confluent Inc.

CLI v4.1.2-SNAPSHOT, Server v4.1.2-SNAPSHOT located at http://ec2-blah.us-blah.compute.amazonaws.com:8080

Having trouble? Type 'help' (case-insensitive) for a rundown of how things work!

ksql>
```

#### Connect to a non-dockerized ksql server running on local host
I haven't managed to get this to work. There are some magic docker incantations I need to learn.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: cp-ksql-cli/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+COPY include/etc/confluent/docker /etc/confluent/docker
+
+RUN mkdir -p /etc/${COMPONENT} /etc/${COMPONENT}/secrets /var/log/${COMPONENT}
+
+RUN bash /etc/confluent/docker/configure
+ENV KSQL_LOG4J_OPTS=-Dlog4j.configuration=file:/etc/${COMPONENT}/log4j.properties
+
+ENTRYPOINT ["ksql"]
+CMD ["http://localhost:8088"]
```
Comment (aayars) [inline cp-ksql-cli/Dockerfile]: This should probably be a real command -- what should the image do when called with no args?

Comment (apurvam) [reply cp-ksql-cli/Dockerfile]: What do you mean by real command? The `ksql` cli command without args will try to connect to the server at `http://localhost:8088`. So technically this extra command is totally redundant.

Not sure if we have too many more options here. The only other option is to not provide any start command, so the user would explicitly have to do `docker run -it -rm confluentinc/cp-ksql-cli ksql <server address>`. 

I think omitting the extra `ksql` command is a bit better, but am open to suggestions.

Comment (aayars) [reply cp-ksql-cli/Dockerfile]: Ah ha.. I didn't realize the behavior of CMD when used with ENTRYPOINT -- I withdraw my comment.

Comment (miguno) [reply cp-ksql-cli/Dockerfile]: Perhaps @gnschenker knows the pros/cons here?

**Comment by miguno** [comment]
@apurvam : Shouldn't this PR (and the server PR) be targeted at `4.1.0-post` and/or `4.1.1-post` rather than 4.1.x?  Because we also want to ship these images for CP 4.1.0/4.1.1?

**Comment by apurvam** [comment]
@miguno

> @apurvam : Shouldn't this PR (and the server PR) be targeted at 4.1.0-post and/or 4.1.1-post rather than 4.1.x? Because we also want to ship these images for CP 4.1.0/4.1.1?

Not sure we can cut these images for the past versions since those tags are already cut. It is dependent on the tooling @aayars is working on, perhaps he can comment?

File: cp-ksql-cli/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
```
Comment (gnschenker) [inline cp-ksql-cli/Dockerfile]: There are too many layers in this image. You should consider to reduce the amount of layers. A lot of layers are causing the graph driver to work too hard and is not recommended for production or production like use.

File: Jenkinsfile [hunk 3]
```diff
@@ -1,9 +1,10 @@
 #!/usr/bin/env groovy
 
 dockerfile {
+    dockerUpstreamTag = '4.1.x-latest'
     slackChannel = '#ksql-eng'
     upstreamProjects = 'confluentinc/schema-registry'
-    dockerRepos = ['confluentinc/ksql-examples', 'confluentinc/ksql-cli', 'confluentinc/ksql-clickstream-demo']
+    dockerRepos = ['confluentinc/ksql-examples', 'confluentinc/ksql-cli', 'confluentinc/ksql-clickstream-demo', 'confluentinc/cp-ksql-cli']
```
Comment (miguno) [inline Jenkinsfile]: Is `cp-ksql-cli` the image name we want?

I am ok with this (and rather keep this name), though @ybyzek advocated to rename from `ksql-cli` to `ksql` in a somewhat related context: https://github.com/confluentinc/ksql/issues/1204.

File: cp-ksql-cli/Dockerfile (hunk 2 above)
Comment (apurvam) [reply cp-ksql-cli/Dockerfile]: What are the excessive layers in your opinion?

File: Jenkinsfile (hunk 3 above)
Comment (apurvam) [reply Jenkinsfile]: I think we should name the new docker image `cp-ksql-cli` for a couple of reasons:

1. The `cp-` prefix keeps it consistent with other docker images provided by Confluent.
2. The `-cli` suffix keeps it consistent with the `-server` suffix for the `cp-ksql-server` docker image.

Comment (miguno) [reply Jenkinsfile]: As I said above, I am +1 regarding `cp-ksql-cli` as the name. Just saying in case it wasn't clear.

**Comment by miguno** [comment]
@apurvam : Why don't we allow, like for the KSQL server image, the configuration of CLI containers via environment variables?

**Comment by ybyzek** [comment]
1. For consistency with the rest of the Docker images, it doesn't seem consistent to ship this with a configuration file instead of ENV variables.

@apurvam have you seen the example in  https://github.com/confluentinc/cp-docker-images/tree/master/debian/kafka/include/etc/confluent/docker  ?  This looks like the model for how to allow the `ksql-cli` to use ENV variables instead of the configuration file.  The change would be to replace

```bash
CMD ["http://localhost:8088"]
```

With 

```bash
CMD ["/etc/confluent/docker/run"]
```

which then implement the configure/ensure/launch steps ( https://github.com/confluentinc/cp-docker-images/blob/master/debian/kafka/include/etc/confluent/docker/run#L38-L45 ).  Then it's the launch script that would call the CLI command.


2. Can you elaborate on "Since we can't log to stdout (that will interfere with the actual CLI), the logs go to /var/logs/ksql-cli/." ?   Is there a way to redirect the output to force the logs to go to Docker logs?  This is also inconsistent with the other Docker images.

**Comment by apurvam** [comment]
@ybyzek , this is not like other docker images: everything else ships servers, not CLIs. 

The problem with emulating the existing `CMD ["/etc/confluent/docker/run"]` ultimately ending up with `exec <service start command>` is that the exec is really final. You cannot append arguments to it. So something like `docker run -it -rm cp-ksql-cli --query-timeout 100 --query-row-limit 100` etc. would not be possible. So we can't both run the script to generate the config file and also the append cli commands (which _can't_ be provided any other way).  

So the existing pattern doesn't work. I'd be happy to take suggestions that do work as a follow up improvement.

Regarding #2, there are no other images which launch an interactive session are there? If there is an example to stay consistent with, we will gladly emulate it. `docker logs` only emits stdout of process id 0. `stdout` for the ksql cli is the actual CLI session. The behavior of docker logs _may_ be configurable, but I couldn't find a way.

**Comment by apurvam** [comment]
Also note that I am only holding of merging this to prevent jeopardizing master build stability while the preview release is being cut. This will be merged as is, the preview images will be cut in this form, and then we can address folllowup later.

**Comment by apurvam** [comment]
@ybyzek 

> .. those parameters are not allowable via properties (ENV) nor in the CLI itself? If yes, it sounds like this requires a product enhancement to KSQL CLI itself, and then the Docker pieces would fall into place.

I don't think this is a product gap. It is natural to have command line options for CLI tools. For instance, if you want to specify a username and password in the future, it would be a bit cumbersome to only have to specify it in some config file. As such, the CLI would support a wide variety cli options which would keep growing. If we go the env variable route, then we close the door to using those options through the docker image AFAICS.

> If the user had to explicitly type ksql, would this resolve the problem such that ENV variables would be usable without a configuration file? 

No. At least not with the existing patterns we use. See [this](https://aboullaite.me/dockerfile-run-vs-cmd-vs-entrypoint/) for a rundown of the `RUN` vs `CMD` vs `ENTRYPOINT`. They key is that if you use `CMD` and then supply another argument on the `docker run` command line, the `CMD` command is never executed. 

In this case, if you specify the `ksql` command as an argument to `docker run`  then the `/etc/confluent/docker/run` script will not be executed, and hence the config file will not be generated anyway. 

As such, anyway you slice it, we have to pick either allowing arguments to the CLI, or allowing configuration through environment variables. We can't have both.

**Comment by ybyzek** [comment]
The product gap isn't that those options are allowed through command line options, it's that they are allowed *only* through command line options.

File: cp-ksql-cli/Dockerfile (hunk 2 above)
Comment (miguno) [reply cp-ksql-cli/Dockerfile]: ping @gnschenker

**Pull request merged**